File: java/org.financer.client.javafx/src/main/java/org/financer/client/javafx/main/profile/ProfileController.java
package org.financer.client.javafx.main.profile;

import com.jfoenix.controls.JFXButton;
import javafx.application.Platform;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.control.*;
import javafx.scene.control.cell.TextFieldTreeCell;
import javafx.util.StringConverter;
import org.controlsfx.glyphfont.FontAwesome;
import org.controlsfx.glyphfont.GlyphFont;
import org.controlsfx.glyphfont.GlyphFontRegistry;
import org.financer.client.domain.api.RestApi;
import org.financer.client.domain.api.RestApiImpl;
import org.financer.client.domain.model.category.Category;
import org.financer.client.domain.model.category.CategoryRoot;
import org.financer.client.domain.model.user.User;
import org.financer.client.format.Formatter;
import org.financer.client.format.FormatterImpl;
import org.financer.client.format.I18N;
import org.financer.client.javafx.dialogs.FinancerConfirmDialog;
import org.financer.client.javafx.dialogs.FinancerTextInputDialog;
import org.financer.client.javafx.local.LocalStorageImpl;
import org.financer.client.javafx.main.FinancerController;
import org.financer.client.local.Application;
import org.financer.client.local.LocalStorage;
import org.financer.shared.domain.model.value.objects.CategoryClass;
import org.financer.util.collections.TreeUtil;

import java.net.URL;
import java.util.ResourceBundle;
import java.util.Set;
import java.util.logging.Logger;

public class ProfileController implements Initializable {

    @FXML
    public Label fullNameLabel;
    @FXML
    public Label emailLabel;
    @FXML
    public Label birthDateLabel;
    @FXML
    public Label addressLabel;
    @FXML
    public Label genderLabel;
    @FXML
    public Hyperlink changePasswordLink;
    @FXML
    public TreeView<Category> categoriesTreeView;
    @FXML
    public JFXButton refreshCategoriesBtn;
    @FXML
    public JFXButton newCategoryBtn;
    @FXML
    public JFXButton editCategoryBtn;
    @FXML
    public JFXButton deleteCategoryBtn;
    @FXML
    public JFXButton editPersonalInformationBtn;

    private final RestApi restApi = new RestApiImpl();
    private final LocalStorage localStorage = LocalStorageImpl.getInstance();
    private final Formatter formatter = new FormatterImpl(localStorage);

    private User user;
    private Logger logger = Logger.getLogger("FinancerApplication");
    private CategoryRoot categoryRoot;
    private TreeItem<Category> treeStructure;

    @Override
    public void initialize(URL location, ResourceBundle resources) {
        FinancerController.setInitializationThread(new Thread(() -> {
            FinancerController.getInstance().showLoadingBox();
            this.user = this.localStorage.readObject("user");
            if (user != null) {
                Platform.runLater(this::fillPersonalInformationLabels);
            }

            this.changePasswordLink.setOnAction(event -> {
                ChangePasswordDialog dialog = new ChangePasswordDialog(user);
                dialog.setOnConfirm(updatedUser -> restApi.updateUsersPassword(user.getId(), updatedUser.getPassword(), result ->
                        FinancerController.getInstance().showToast(Application.MessageType.SUCCESS, I18N.get("succChangedPassword"))).execute());
            });

            restApi.updateUsersPersonalInformation(user, result -> {
                FinancerController.getInstance().showToast(Application.MessageType.SUCCESS, I18N.get("succChangedPersonalInformation"));
                localStorage.writeObject("user", user);
            }).execute();

            GlyphFont fontAwesome = GlyphFontRegistry.font("FontAwesome");
            this.refreshCategoriesBtn.setGraphic(fontAwesome.create(FontAwesome.Glyph.REFRESH));
            this.refreshCategoriesBtn.setGraphicTextGap(8);
            this.newCategoryBtn.setGraphic(fontAwesome.create(FontAwesome.Glyph.PLUS));
            this.newCategoryBtn.setGraphicTextGap(8);
            this.newCategoryBtn.setDisable(true);
            this.editCategoryBtn.setGraphic(fontAwesome.create(FontAwesome.Glyph.EDIT));
            this.editCategoryBtn.setGraphicTextGap(8);
            this.editCategoryBtn.setDisable(true);
            this.deleteCategoryBtn.setGraphic(fontAwesome.create(FontAwesome.Glyph.TRASH));
            this.deleteCategoryBtn.setGraphicTextGap(8);
            this.deleteCategoryBtn.setDisable(true);

            this.handleRefreshCategories();

            this.editPersonalInformationBtn.setOnAction(event -> {
                ChangePersonalInformationDialog dialog = new ChangePersonalInformationDialog(this.user);
                dialog.setOnConfirm(user -> {
                    if (user != null) {
                        restApi.updateUsersPersonalInformation(user, result -> {
                            FinancerController.getInstance().showToast(Application.MessageType.SUCCESS, I18N.get("succChangedPersonalInformation"));
                            localStorage.writeObject("user", user);
                            Platform.runLater(this::fillPersonalInformationLabels);
                        }).execute();
                    }
                });
            });
            this.newCategoryBtn.setDisable(true);
            this.editCategoryBtn.setDisable(true);
            this.deleteCategoryBtn.setDisable(true);

            categoryRoot = localStorage.readObject("categories");
            this.loadCategoryData();
            FinancerController.getInstance().hideLoadingBox();
        }));
        FinancerController.getInitializationThread().start();
    }

    private void fillPersonalInformationLabels() {
        this.fullNameLabel.setText(formatter.format(user.getName()));
        this.emailLabel.setText(user.getEmail().getEmailAddress());
        this.birthDateLabel.setText(formatter.format(user.getBirthDate()));
        this.genderLabel.setText(I18N.get(this.user.getGender().getGender().getName()));
    }

    public void handleRefreshCategories() {
        restApi.getUsersCategories(user.getId(), result -> {
            categoryRoot = new CategoryRoot(result);
            categoryRoot = localStorage.readObject("categories");
            Platform.runLater(this::loadCategoryData);
        }).execute();
    }

    private void loadCategoryData() {
        createTreeView();
        categoriesTreeView.getSelectionModel().selectedItemProperty().addListener((observableValue, oldValue, newValue) -> {
            if (newValue != null) {
                newCategoryBtn.setDisable(false);

                editCategoryBtn.setDisable(newValue.getValue().isRoot());
                deleteCategoryBtn.setDisable(newValue.getValue().isRoot());
            }
        });
        categoriesTreeView.setEditable(false);
        categoriesTreeView.setShowRoot(false);
        Platform.runLater(() -> categoriesTreeView.setRoot(treeStructure));
        expandTreeView(treeStructure);
        categoriesTreeView.setCellFactory(param -> getCellFactory());
        categoriesTreeView.setOnEditCommit(event -> {
            event.getNewValue().setId(event.getOldValue().getId());
            event.getNewValue().setParent(event.getOldValue().getParent());
            event.getNewValue().setCategoryClass(event.getOldValue().getCategoryClass());
            handleUpdateCategory(event.getNewValue());
        });
    }

    private void initializePersonalInformation() {
        this.fullNameLabel.setText(formatter.format(user.getName()));
        this.emailLabel.setText(user.getEmail().getEmailAddress());
        this.birthDateLabel.setText(formatter.format(user.getBirthDate()));
        this.genderLabel.setText(I18N.get(this.user.getGender().getGender().getName()));
    }

    public void handleNewCategory() {
        this.handleNewCategory(this.categoriesTreeView.getSelectionModel().getSelectedItem());
    }

    private void handleNewCategory(TreeItem<Category> currentItem) {
        if (currentItem != null) {
            FinancerTextInputDialog dialog = new FinancerTextInputDialog(I18N.get("enterCategoryName"), I18N.get("newCategory"));
            dialog.setOnConfirm(categoryName -> {
                Category newCategory = new Category()
                        .setParent(currentItem.getValue())
                        .setCategoryClass(currentItem.getValue().getCategoryClass())
                        .setName(categoryName);

                restApi.createCategory(newCategory, result -> {
                    currentItem.getChildren().add(new TreeItem<>(result));

                    if (categoriesTreeView.getSelectionModel().getSelectedItem().getValue().isRoot()) {
                        categoryRoot.getCategoriesByClass(newCategory.getCategoryClass()).add(newCategory);
                    } else {
                        TreeUtil.insertByValue(categoryRoot, newCategory, (o1, o2) -> Long.compare(((Category) o1).getParent().getId(), ((Category) o2).getId()));
                    }
                    localStorage.writeObject("categories", categoryRoot);

                    Platform.runLater(() -> {
                        categoriesTreeView.getSelectionModel().getSelectedItem().setExpanded(true);
                        categoriesTreeView.refresh();
                    });
                }).execute();
            });
        }
    }

    public void handleEditCategory() {
        if (this.categoriesTreeView.getSelectionModel().getSelectedItem() != null) {
            Category category = this.categoriesTreeView.getSelectionModel().getSelectedItem().getValue();
            FinancerTextInputDialog dialog = new FinancerTextInputDialog(I18N.get("enterCategoryName"), category.getName());
            dialog.setOnConfirm(categoryName -> {
                category.setName(categoryName);
                this.categoriesTreeView.refresh();
                this.handleUpdateCategory(category);
            });
        }
    }

    private void handleUpdateCategory(Category category) {
        restApi.updateCategory(category, result -> {
            categoriesTreeView.getSelectionModel().getSelectedItem().getValue().setName(category.getName());
            localStorage.writeObject("categories", categoryRoot);
        }).execute();
    }

    public void handleDeleteCategory() {
        this.handleDeleteCategory(this.categoriesTreeView.getSelectionModel().getSelectedItem().getValue());
    }

    private void handleDeleteCategory(Category category) {
        if (category != null && !category.isRoot()) {
            FinancerConfirmDialog dialog = new FinancerConfirmDialog(I18N.get("confirmDeleteCategory"));
            dialog.setOnConfirm(result -> restApi.deleteCategory(category.getId(), voidResult -> {
                TreeUtil.deleteByValue(categoryRoot, categoriesTreeView.getSelectionModel().getSelectedItem().getValue());
                localStorage.writeObject("categories", categoryRoot);
                Platform.runLater(() -> categoriesTreeView.getSelectionModel().getSelectedItem().getParent().getChildren().remove(categoriesTreeView.getSelectionModel().getSelectedItem()));
            }).execute());
        }
    }

    private void createTreeView() {
        this.treeStructure = new TreeItem<>(new Category().setName("root"));
        for (CategoryClass categoryClass : CategoryClass.getAll()) {
            Set<Category> categories = this.categoryRoot.getCategoriesByClass(categoryClass);
            for (Category category : categories) {
                this.createTreeView(treeStructure, category);
            }
        }
    }

    private void createTreeView(TreeItem<Category> treeItem, Category root) {
        TreeItem<Category> treeRoot = new TreeItem<>(root);
        treeItem.getChildren().add(treeRoot);
        if (!root.isLeaf()) {
            for (Category child : root.getChildren()) {
                createTreeView(treeRoot, child);
            }
        }
    }

    private TextFieldTreeCellImpl getCellFactory() {
        return new TextFieldTreeCellImpl(new StringConverter<>() {
            @Override
            public String toString(Category category) {
                return category.getName().equals(category.getCategoryClass().getCategoryClass().getName()) ? I18N.get(category.getName()) : category.getName();
            }

            @Override
            public Category fromString(String string) {
                return new Category().setName(string);
            }
        });
    }

    private void expandTreeView(TreeItem<?> item) {
        if (item != null && !item.isLeaf()) {
            item.setExpanded(true);
            for (TreeItem<?> child : item.getChildren()) {
                expandTreeView(child);
            }
        }
    }

    private final class TextFieldTreeCellImpl extends TextFieldTreeCell<Category> {
        private final ContextMenu contextMenu = new ContextMenu();
        private final ContextMenu deleteContextMenu = new ContextMenu();

        TextFieldTreeCellImpl(StringConverter<Category> stringConverter) {
            super(stringConverter);

            MenuItem addMenuItem = new MenuItem(I18N.get("new"));
            addMenuItem.setOnAction(t -> handleNewCategory());
            this.contextMenu.getItems().add(addMenuItem);

            MenuItem addMenuItemDelete = new MenuItem(I18N.get("new"));
            addMenuItemDelete.setOnAction(t -> getTreeItem().getChildren().add(new TreeItem<>(new Category().setName("newCategory"))));
            this.deleteContextMenu.getItems().add(addMenuItemDelete);

            MenuItem deleteMenuItem = new MenuItem(I18N.get("delete"));
            this.deleteContextMenu.getItems().add(deleteMenuItem);
            deleteMenuItem.setOnAction(t -> handleDeleteCategory(getTreeItem().getValue()));
        }

        @Override
        public void updateItem(Category item, boolean empty) {
            super.updateItem(item, empty);
            if (item != null && !isEditing() && getParent() != null) {
                if (item.isRoot() && !formatter.format(item).equals(I18N.get("categories"))) {
                    setContextMenu(this.contextMenu);
                } else if (!item.isRoot()) {
                    setContextMenu(this.deleteContextMenu);
                }
            }
        }
    }
}

File: java/org.financer.client.javafx/src/main/kotlin/org/financer/client/javafx/main/profile/ProfileController.kt
package org.financer.client.javafx.main.profile

import com.jfoenix.controls.JFXButton
import javafx.application.Platform
import javafx.beans.value.ObservableValue
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.scene.control.*
import javafx.scene.control.cell.TextFieldTreeCell
import javafx.util.StringConverter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.controlsfx.glyphfont.FontAwesome
import org.controlsfx.glyphfont.GlyphFontRegistry
import org.financer.client.domain.api.RestApi
import org.financer.client.domain.api.RestApiImpl
import org.financer.client.domain.model.category.Category
import org.financer.client.domain.model.category.CategoryRoot
import org.financer.client.domain.model.user.User
import org.financer.client.format.Formatter
import org.financer.client.format.FormatterImpl
import org.financer.client.format.I18N.get
import org.financer.client.javafx.dialogs.FinancerConfirmDialog
import org.financer.client.javafx.dialogs.FinancerTextInputDialog
import org.financer.client.javafx.local.LocalStorageImpl
import org.financer.client.javafx.main.FinancerController
import org.financer.client.local.Application
import org.financer.shared.domain.model.value.objects.CategoryClass
import org.financer.util.collections.Action
import org.financer.util.collections.Tree
import org.financer.util.collections.TreeUtil
import java.net.URL
import java.util.*
import java.util.logging.Logger

class ProfileController : Initializable {
    @FXML
    var fullNameLabel: Label? = null

    @FXML
    var emailLabel: Label? = null

    @FXML
    var birthDateLabel: Label? = null

    @FXML
    var addressLabel: Label? = null

    @FXML
    var genderLabel: Label? = null

    @FXML
    var changePasswordLink: Hyperlink? = null

    @FXML
    var categoriesTreeView: TreeView<Category>? = null

    @FXML
    var refreshCategoriesBtn: JFXButton? = null

    @FXML
    var newCategoryBtn: JFXButton? = null

    @FXML
    var editCategoryBtn: JFXButton? = null

    @FXML
    var deleteCategoryBtn: JFXButton? = null

    @FXML
    var editPersonalInformationBtn: JFXButton? = null
    private val restApi: RestApi = RestApiImpl()
    private val localStorage = LocalStorageImpl.getInstance()
    private val formatter: Formatter = FormatterImpl(localStorage)
    private lateinit var user: User
    private val logger = Logger.getLogger("FinancerApplication")
    private lateinit var categoryRoot: CategoryRoot
    private var treeStructure: TreeItem<Category>? = null
    override fun initialize(location: URL, resources: ResourceBundle) {
        CoroutineScope(Dispatchers.Default).launch {
            FinancerController.instance!!.showLoadingBox()
            user = localStorage.readObject("user") ?: throw IllegalStateException("No User stored in local storage.")
            withContext(Dispatchers.Main) { fillPersonalInformationLabels() }
            changePasswordLink!!.onAction = EventHandler {
                val dialog = ChangePasswordDialog(user)
                dialog.setOnConfirm(Action { updatedUser: User? ->
                    CoroutineScope(Dispatchers.IO).launch {
                        restApi.updateUsersPassword(user.getId(), updatedUser!!.password!!)
                        FinancerController.instance!!.showToast(Application.MessageType.SUCCESS, get("succChangedPassword"))
                    }
                })
            }

            CoroutineScope(Dispatchers.IO).launch {
                val updateUser = restApi.updateUsersPersonalInformation(user)
                if (updateUser != null) {
                    FinancerController.instance!!.showToast(Application.MessageType.SUCCESS, get("succChangedPersonalInformation"))
                    user = updateUser
                    localStorage.writeObject("user", updateUser)
                }
            }
            val fontAwesome = GlyphFontRegistry.font("FontAwesome")
            refreshCategoriesBtn!!.graphic = fontAwesome.create(FontAwesome.Glyph.REFRESH)
            refreshCategoriesBtn!!.graphicTextGap = 8.0
            newCategoryBtn!!.graphic = fontAwesome.create(FontAwesome.Glyph.PLUS)
            newCategoryBtn!!.graphicTextGap = 8.0
            newCategoryBtn!!.isDisable = true
            editCategoryBtn!!.graphic = fontAwesome.create(FontAwesome.Glyph.EDIT)
            editCategoryBtn!!.graphicTextGap = 8.0
            editCategoryBtn!!.isDisable = true
            deleteCategoryBtn!!.graphic = fontAwesome.create(FontAwesome.Glyph.TRASH)
            deleteCategoryBtn!!.graphicTextGap = 8.0
            deleteCategoryBtn!!.isDisable = true
            handleRefreshCategories()
            editPersonalInformationBtn!!.onAction = EventHandler { event: ActionEvent? ->
                val dialog = ChangePersonalInformationDialog(user)
                dialog.setOnConfirm(Action { user: User? ->
                    if (user != null) {
                        CoroutineScope(Dispatchers.IO).launch {
                            val updatedUser = restApi.updateUsersPersonalInformation(user)
                            if (updatedUser != null) {
                                FinancerController.instance!!.showToast(Application.MessageType.SUCCESS, get("succChangedPersonalInformation"))
                                localStorage.writeObject("user", updatedUser)
                                withContext(Dispatchers.Main) { fillPersonalInformationLabels() }
                            }
                        }
                    }
                })
            }
            newCategoryBtn!!.isDisable = true
            editCategoryBtn!!.isDisable = true
            deleteCategoryBtn!!.isDisable = true
            categoryRoot = localStorage.readObject("categories")
                    ?: throw IllegalStateException("No categories stored in local storage.");
            loadCategoryData()
            FinancerController.instance!!.hideLoadingBox()
        }
    }

    private fun fillPersonalInformationLabels() {
        fullNameLabel!!.text = formatter.format(user.name!!)
        emailLabel!!.text = user.email!!.emailAddress
        birthDateLabel!!.text = formatter.format(user.birthDate!!)
        genderLabel!!.text = get(user.gender!!.gender.getName())
    }

    fun handleRefreshCategories() = CoroutineScope(Dispatchers.IO).launch {
        val result = restApi.getUsersCategories(user.getId())
        categoryRoot = CategoryRoot(result.toSet())
        localStorage.writeObject("categories", categoryRoot)
        withContext(Dispatchers.Main) { loadCategoryData() }
    }

    private fun loadCategoryData() {
        createTreeView()
        categoriesTreeView!!.selectionModel.selectedItemProperty().addListener { observableValue: ObservableValue<out TreeItem<Category>?>?, oldValue: TreeItem<Category>?, newValue: TreeItem<Category>? ->
            if (newValue != null) {
                newCategoryBtn!!.isDisable = false
                editCategoryBtn!!.isDisable = newValue.value.isRoot
                deleteCategoryBtn!!.isDisable = newValue.value.isRoot
            }
        }
        categoriesTreeView!!.isEditable = false
        categoriesTreeView!!.isShowRoot = false
        Platform.runLater { categoriesTreeView!!.setRoot(treeStructure) }
        expandTreeView(treeStructure)
        categoriesTreeView!!.setCellFactory { param: TreeView<Category>? -> cellFactory }
        categoriesTreeView!!.setOnEditCommit { event: TreeView.EditEvent<Category> ->
            event.newValue.setId(event.oldValue.id)
            event.newValue.parent = event.oldValue.parent
            event.newValue.setCategoryClass(event.oldValue.categoryClass)
            handleUpdateCategory(event.newValue)
        }
    }

    fun handleNewCategory() {
        this.handleNewCategory(categoriesTreeView!!.selectionModel.selectedItem)
    }

    private fun handleNewCategory(currentItem: TreeItem<Category>?) {
        if (currentItem != null) {
            val dialog = FinancerTextInputDialog(get("enterCategoryName"), get("newCategory"))
            dialog.setOnConfirm(Action { categoryName: String? ->
                val newCategory = Category()
                        .setParent(currentItem.value)
                        .setCategoryClass(currentItem.value.categoryClass)
                        .setName(categoryName)
                CoroutineScope(Dispatchers.IO).launch {
                    val result = restApi.createCategory(newCategory)
                    if (result != null) {
                        currentItem.children.add(TreeItem(result))
                        if (categoriesTreeView!!.selectionModel.selectedItem.value.isRoot) {
                            categoryRoot.getCategoriesByClass(newCategory.categoryClass!!).add(newCategory)
                        } else {
                            TreeUtil.insertByValue(categoryRoot, newCategory) { o1: Tree, o2: Tree -> (o1 as Category).parent.id.compareTo((o2 as Category).id) }
                        }
                        localStorage.writeObject("categories", categoryRoot)
                        Platform.runLater {
                            categoriesTreeView!!.selectionModel.selectedItem.isExpanded = true
                            categoriesTreeView!!.refresh()
                        }
                    }

                }
            })
        }
    }

    fun handleEditCategory() {
        if (categoriesTreeView!!.selectionModel.selectedItem != null) {
            val category = categoriesTreeView!!.selectionModel.selectedItem.value
            val dialog = FinancerTextInputDialog(get("enterCategoryName"), category.name)
            dialog.setOnConfirm(Action { categoryName: String? ->
                category.setName(categoryName)
                categoriesTreeView!!.refresh()
                handleUpdateCategory(category)
            })
        }
    }

    private fun handleUpdateCategory(category: Category) = CoroutineScope(Dispatchers.IO).launch {
        restApi.updateCategory(category)
        categoriesTreeView!!.selectionModel.selectedItem.value.setName(category.name)
        localStorage.writeObject("categories", categoryRoot)
    }

    fun handleDeleteCategory() {
        this.handleDeleteCategory(categoriesTreeView!!.selectionModel.selectedItem.value)
    }

    private fun handleDeleteCategory(category: Category?) {
        if (category != null && !category.isRoot) {
            val dialog = FinancerConfirmDialog(get("confirmDeleteCategory"))
            dialog.setOnConfirm(Action {
                CoroutineScope(Dispatchers.IO).launch {
                    restApi.deleteCategory(category.id)
                    TreeUtil.deleteByValue(categoryRoot, categoriesTreeView!!.selectionModel.selectedItem.value)
                    localStorage.writeObject("categories", categoryRoot)
                    Platform.runLater { categoriesTreeView!!.selectionModel.selectedItem.parent.children.remove(categoriesTreeView!!.selectionModel.selectedItem) }
                }
            })
        }
    }

    private fun createTreeView() {
        treeStructure = TreeItem(Category().setName("root"))
        for (categoryClass in CategoryClass.getAll()) {
            val categories = categoryRoot.getCategoriesByClass(categoryClass)
            for (category in categories) {
                this.createTreeView(treeStructure!!, category)
            }
        }
    }

    private fun createTreeView(treeItem: TreeItem<Category>, root: Category) {
        val treeRoot = TreeItem(root)
        treeItem.children.add(treeRoot)
        if (!root.isLeaf) {
            for (child in root.children) {
                createTreeView(treeRoot, child)
            }
        }
    }

    private val cellFactory: TextFieldTreeCellImpl = TextFieldTreeCellImpl(object : StringConverter<Category>() {
        override fun toString(category: Category): String {
            return (if (category.name == category.categoryClass!!.categoryClass.getName()) get(category.name) else category.name)!!
        }

        override fun fromString(string: String): Category? {
            return Category().setName(string)
        }
    })

    private fun expandTreeView(item: TreeItem<*>?) {
        if (item != null && !item.isLeaf) {
            item.isExpanded = true
            for (child in item.children) {
                expandTreeView(child)
            }
        }
    }

    private inner class TextFieldTreeCellImpl internal constructor(stringConverter: StringConverter<Category>)
        : TextFieldTreeCell<Category>(stringConverter) {
        private val customContextMenu = ContextMenu()
        private val deleteContextMenu = ContextMenu()
        override fun updateItem(item: Category?, empty: Boolean) {
            super.updateItem(item, empty)
            if (item != null && !isEditing && parent != null) {
                if (item.isRoot && formatter.format(item) != get("categories")) {
                    contextMenu = this.customContextMenu
                } else if (!item.isRoot) {
                    contextMenu = deleteContextMenu
                }
            }
        }

        init {
            val addMenuItem = MenuItem(get("new"))
            addMenuItem.onAction = EventHandler { t: ActionEvent? -> handleNewCategory() }
            this.customContextMenu.items.add(addMenuItem)
            val addMenuItemDelete = MenuItem(get("new"))
            addMenuItemDelete.onAction = EventHandler { t: ActionEvent? -> treeItem.children.add(TreeItem(Category().setName("newCategory"))) }
            deleteContextMenu.items.add(addMenuItemDelete)
            val deleteMenuItem = MenuItem(get("delete"))
            deleteContextMenu.items.add(deleteMenuItem)
            deleteMenuItem.onAction = EventHandler { t: ActionEvent? -> handleDeleteCategory(treeItem.value) }
        }
    }
}